crates/bin/prove_block/src/main.rs

```diff
@@ -0,0 +1,741 @@
+
+                        compiled_classes.insert(class_hash, casm_cc);
+                    }
+                    starknet::core::types::ContractClass::Legacy(_compressed_logacy_contract_class) => {
+                        panic!("legacy class (TODO)");
+                    }
+                }
+            } else {
+                log::warn!("No class available for contract {}", contract_address);
+            }
+        } else {
+            log::warn!("No class hash available for contract {}", contract_address);
+        };
+    }
+
+    let blockifier_state_reader = RpcStateReader { block_id: BlockId::Number(block_number - 1), rpc_client: provider };
+
+    let tx_execution_infos = reexecute_transactions_with_blockifier(
+        CachedState::new(blockifier_state_reader, GlobalContractCache::new(1024)),
+        &block_context,
+        block_with_txs.transactions.iter().map(|tx| starknet_rs_to_blockifier(tx).unwrap()).collect(),
+    )?;
+
+    if tx_execution_infos.len() != transactions.len() {
+        log::warn!(
+            "Warning: blockifier reexecution yielded different num execution infos ({}) than transactions ({})",
+            tx_execution_infos.len(),
+            transactions.len()
+        );
+    }
+
+    let contract_state_commitment_info =
+        CommitmentInfo::create_from_modifications::<CachedRpcStorage, PedersenHash, ContractState>(
+            previous_tree.clone(),
+            None, // TODO: do we have a source for expected?
+            updates,
+            &mut initial_state.ffc,
+        )
+        .await?;
+
+    let os_input = StarknetOsInput {
+        contract_state_commitment_info,
+        contract_class_commitment_info: Default::default(),
+        deprecated_compiled_classes: Default::default(),
+        compiled_classes,
+        compiled_class_visited_pcs: Default::default(),
+        contracts: contract_states,
+        class_hash_to_compiled_class_hash,
+        general_config,
+        transactions,
+        block_hash: block_with_txs.block_hash,
+    };
+    let execution_helper = ExecutionHelperWrapper::<CachedRpcStorage>::new(
+        contract_storages,
+        tx_execution_infos,
+        &block_context,
+        (old_block_number, old_block_hash),
+    );
+
+    let result = run_os(config::DEFAULT_COMPILED_OS, layout, os_input, block_context, execution_helper);
+
+    match &result {
+        Err(Runner(VmException(vme))) => {
+            if let Some(traceback) = vme.traceback.as_ref() {
+                log::error!("traceback:\n{}", traceback);
+            }
+            if let Some(inst_location) = &vme.inst_location {
+                log::error!("died at: {}:{}", inst_location.input_file.filename, inst_location.start_line);
+                log::error!("inst_location:\n{:?}", inst_location);
+            }
+        }
+        Err(_) => {
+            println!("exception:\n{:#?}", result);
+        }
+        _ => {}
+    }
+
+    result.unwrap();
+
+    Ok(())
```

Comment:
This monster of a function needs to be split into several smaller functions to be reviewable.